distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Common/Encoding_Limited_Naming_Properties.enso

```diff
@@ -0,0 +1,63 @@
+from Standard.Base import all
+import Standard.Base.Errors.Encoding_Error.Encoding_Error
+import Standard.Base.Errors.Illegal_Argument.Illegal_Argument
+import Standard.Base.Errors.Illegal_State.Illegal_State
+
+from Standard.Table.Errors import Name_Too_Long
+
+## PRIVATE
+   A `Naming_Properties` implementation which sets the size limit based on the
+   size of a string encoded in a particular encoding, in bytes.
+type Encoding_Limited_Naming_Properties
+    ## PRIVATE
+    Instance (encoding : Encoding) (limit : Integer)
+
+    ## PRIVATE
+    encoded_size : Text -> Integer
+    encoded_size self name =
+        r = name.bytes encoding=self.encoding on_problems=Problem_Behavior.Report_Error . length
+        r.catch Encoding_Error error->
+            Panic.throw (Illegal_Argument.Error "The name ["+name.to_display_text+"] contains characters that could not be encoded in the database server-size encoding "+self.encoding.to_display_text+"." cause=error)
+
+    ## PRIVATE
+    truncate : Text -> Integer -> Text
+    truncate self name size =
+        ## I assume that `self.encoded_size` is a homomorphism
+           from text with concatenation to integers with addition.
+           I.e. for every a and b, `self.encoded_size (a + b) == self.encoded_size a + self.encoded_size b`.
+        clusters = name.characters
+        if clusters.is_empty then name else
+            cluster_sizes = clusters.map self.encoded_size
+            # We see how many clusters we can fit until we exceed the size.
+            last_cluster_that_still_fits = cluster_sizes.running_fold 0 (+) . last_index_of (<= size)
+            case last_cluster_that_still_fits of
+                # In the very rare edge case where even the first grapheme cluster cannot fit, we have to just return an empty string.
```

Comment:
This edge case would be (1) a very long first character and (2) a very small column/table name restriction; I think this is unlikely enough that it could just be a Panic here.

Reply:
Unfortunately it is slightly more often. The size limit is driven not only by the DB limit but also by the length of the numeric suffix added. So if we need to add a lot of numbers it can grow. Of course its size grows logarithmically so it should be still rare, but within realm of possibility.

On the other hand, returning an empty value may indeed turn out to be problematic - as I'm not sure all code will be able to handle it right. Maybe indeed it will be safer to panic in this case - indeed it should not happen in 'normal' situations.

Reply:
After further thought I think it's safer to leave it not panic for now. I don't think even unhandling the emtpy name case could cause data corruption, worst case scenario it can generate invalid SQL and fail there. But in most cases I imagine it should actually be able to proceed in which case I think that's better than panicking unnecessarily.

Truncating to empty length is not quite a good thing, but it's not an unrecoverable error either, I think.